go-controller/pkg/ovn/libovsdbops/util.go

```diff
@@ -2,12 +2,394 @@ package libovsdbops
 
 import (
 	"context"
+	"errors"
 	"fmt"
+	"math/rand"
+	"reflect"
+	"strings"
+	"time"
 
+	"github.com/mitchellh/copystructure"
 	"github.com/ovn-org/libovsdb/client"
+	"github.com/ovn-org/libovsdb/model"
 	"github.com/ovn-org/libovsdb/ovsdb"
+	"k8s.io/klog"
 )
 
+var (
+	UUIDField = "UUID"
+)
+
+type ModelClient struct {
+	client client.Client
+}
+
+func NewModelClient(client client.Client) ModelClient {
+	return ModelClient{client: client}
+}
+
+/*
+ OnReferentialModelMutation is a helper function which assigns or unassigns
+ all UUIDs provided by `items` against the referential (parent) model.
+
+ Note the following:
+
+- items can be a slice or a unique object. The reasons for this are: a slice
+  makes more sense for delete operations where the client needs to bulk delete a
+  bunch of objects at once (this is done in egressip.go / egressfirewall.go when
+  syncing for example). For create operations usually only one object is to be
+  specified, though multiple can.
+
+- it compares existing referential UUIDs to the ones provided through items, as
+  to avoid round-trips to the OVS DB and unnecessary operations
+*/
+func OnReferentialModelMutation(field interface{}, mutator ovsdb.Mutator, items interface{}) []model.Mutation {
+	ids := []string{}
+	existingIds, ok := field.(*[]string)
+	if !ok {
+		return nil
+	}
+	if reflect.TypeOf(items).Kind() == reflect.Slice {
+		itemsLength := reflect.ValueOf(items).Len()
+		if itemsLength == 0 {
+			return nil
+		}
+		for i := 0; i < itemsLength; i++ {
+			uuidField := reflect.ValueOf(items).Index(i).FieldByName(UUIDField)
+			if uuidField.IsValid() {
+				uuid := uuidField.Interface().(string)
+				if uuid != "" && !SliceHasStringItem(*existingIds, uuid) {
```

Comment:
If this is used for deletes, should the check be the opposite here, `SliceHasStringItem`vs `!SliceHasStringItem` ?

Reply:
Indeed, thanks for spotting that!